*: use cluster-aware clients, listers and informers

stevekuznetsov opened the pull request:
Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

Part of https://github.com/kcp-dev/kcp/issues/1943

stevekuznetsov committed:
UPSTREAM: <carry>: go.mod: add and pin kcp-dev/{apimachinery,client-go}

hack/pin-dependency.sh github.com/kcp-dev/apimachinery 9e1e13940519ce025808e80130b736103dbed45b
hack/pin-dependency.sh github.com/kcp-dev/client-go 387bce248341f5e66ec779d05952fd198c777502
hack/update-vendor.sh

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <carry>: turn off AP&F

We previously had a hack to make this cluster-aware & do nothing in the
root cluster, but there's no sense in that. Just turn it off.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <carry>: add hacks to alloww sneaking past plumbing

We need to provide cluster-aware clients and informer factories to
components by using the k8s plumbing, but actually refactoriung the
types would be a nightmare and require lots of fan-out, so we masquerade
instead.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <squash>: stop providing a formatting helper

The ToClusterAwareKey is being used in two spots:

  1. in accessing listers, to be refactored in a future commit
  2. in formatting queue keys, and other caches, and what-not, rmeoved
     here and replaced with a simple string concatenation, to be removed
     wholesale in the future once we're done

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <squash>: use the sneaky plumbing hacks

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <drop>: use legacy keyfunc in quota controller

These controllers are cluster-unawara, and we run one per workspace, so
they need to queue & pop cluster-unaware keys. We can remove the Legacy*
versions of these keyfuncs once everyone stops relying on the hacked
MetaNamespaceKeyFunc (once all client-sets are moved over)

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <carry>: remove unused code

These code paths were not utilized, or were broken, and rather than
making them compile, I just removed them as they had no function.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <carry>: genericcontrolplane: return informers

We were creating two sets of clients and informer factories before, and
one of them was never even checked to sync. Return the one we create and
use it.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <carry>: admission: encode cluster in attributes

We need to do cluster-specific things deep in this call-stack, and
encoding the cluster in the attributes record is by far the simplest way
to do that.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <carry>: other misc plumbing refactors

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <carry>: make a cluster-aware CRD-aware CRD lister

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <carry>: apiserver: make the token-getter cluster-aware

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <carry>: remove pod lister usgae from token-getter

This was using the duplicate informer factory before, which never
synced, since there are no Pods. In lieu of a bigger refactor, just stop
trying to use the thing.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <carry>: serviceaccounts: use cluster-aware RV cache

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <carry>: expose nicer constructors for admission

We don't have a scoped informer factory itself, so just pass in the
informer.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <carry>: *: semi-automated refactor of clients & listers

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

ncdc commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go:
```diff
@@ -111,24 +113,24 @@ func (a *Webhook) SetServiceResolver(sr webhookutil.ServiceResolver) {
-	if a.hookSource == nil {
-		return fmt.Errorf("kubernetes client is not properly setup")
-	}
+	//if a.hookSource == nil {
```
Do you want to keep this commented out or delete it?

stevekuznetsov replied:
We can delete, yeah

ncdc commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go:
```diff
@@ -47,10 +48,16 @@ func (r *versionDiscoveryHandler) ServeHTTP(w http.ResponseWriter, req *http.Req
 	requestedVersion := pathParts[2]
 
-	crds, err := r.crdLister.List(req.Context(), labels.Everything())
+	crds, err := r.crdLister.Cluster(clusterName).List(req.Context(), labels.Everything())
```
Doing this means we can eventually remove ctx from List() if we feel like it. That's the only reason it's there for now.

stevekuznetsov replied:
Amazing! Yes, we certainly can follow-up with this.

stevekuznetsov committed:
UPSTREAM: <squash>: remove the cluster-aware webhook type

This was never used, so is unnecessary?

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
UPSTREAM: <squash>: genericcontrolplane: don't register ap&f

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

stevekuznetsov committed:
go.mod: bump kcp-dev/client-go

hack/pin-dependency.sh github.com/kcp-dev/client-go 60e56386a574f8c4589b3a6585901b16ed8fc359
hack/update-vendor.sh

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>
